subxt/src/rpc/lightclient/background.rs

```diff
@@ -0,0 +1,301 @@
+use futures::stream::StreamExt;
+use futures_util::future::{self, Either};
+use serde::Deserialize;
+use serde_json::value::RawValue;
+use std::{
+    collections::{hash_map::Entry, HashMap},
+    str::FromStr,
+};
+use tokio::sync::{mpsc, oneshot};
+
+///! The background task of the light client.
+
+/// The number of notifications that are buffered, before the user
+/// registers to [`LightClientInner::register_subscription`].
+/// Only the first `BUFFER_NUM_NOTIFICATIONS` are buffered, while the
+/// others are ignored.
+///
+/// In practice, the Light Client produces notifications at a lower rate
+/// than the substrate full node.
+const BUFFER_NUM_NOTIFICATIONS: usize = 16;
+
+const LOG_TARGET: &str = "light-client-background";
+
+/// Message protocol between the front-end client that submits the RPC requests
+/// and the backend handler that produces responses from the chain.
+///
+/// The light client uses a single object [`smoldot_light::JsonRpcResponses`] to
+/// handle all requests and subscriptions from a chain. A background task is spawned
```

Comment:
That's correct, however "a chain" is actually reference-counted in smoldot's internals. If you add two chains with the same chain spec, it will only connect once and sync once. Rather than multiplex the requests, I would suggest you create multiple "chains", one per user/client/whatever. I can also add a `clone_chain` function to make this easier.